pkg/collector/scheduler/job.go

```diff
@@ -16,91 +17,262 @@ import (
-	jobs     []check.Check
-	running  bool
-	health   *health.Handle
-	mu       sync.RWMutex // to protect critical sections in struct's fields
+	interval        time.Duration
+	stop            chan bool // to stop this queue
+	stopped         chan bool // signals that this queue has stopped
+	buckets         []*jobBucket
+	nextBucket      uint
+	bucketScheduled uint
+	running         bool
+	health          *health.Handle
+	mu              sync.RWMutex // to protect critical sections in struct's fields
 }
 
 // newJobQueue creates a new jobQueue instance
 func newJobQueue(interval time.Duration) *jobQueue {
-	return &jobQueue{
+	jq := &jobQueue{
 		interval: interval,
-		ticker:   time.NewTicker(time.Duration(interval)),
 		stop:     make(chan bool),
 		stopped:  make(chan bool),
 		health:   health.Register("collector-queue"),
 	}
 
+	var nb int
+	if interval <= time.Second {
+		nb = 1
+	} else {
+		nb = int(interval.Truncate(time.Second).Seconds())
+	}
+	for i := 0; i < nb; i++ {
+		bucket := newJobBucket(uint(i), interval, time.Duration(time.Duration(i)*time.Second))
+		jq.buckets = append(jq.buckets, bucket)
+	}
+
+	return jq
 }
 
 // addJob is a convenience method to add a check to a queue
 func (jq *jobQueue) addJob(c check.Check) {
 	jq.mu.Lock()
 	defer jq.mu.Unlock()
 
-	jq.jobs = append(jq.jobs, c)
+	// Checks scheduled to buckets scheduled in round-robin
+	jq.buckets[jq.nextBucket].addJob(c)
+	jq.nextBucket = (jq.nextBucket + 1) % uint(len(jq.buckets))
 }
 
 func (jq *jobQueue) removeJob(id check.ID) error {
 	jq.mu.Lock()
 	defer jq.mu.Unlock()
 
-	for i, c := range jq.jobs {
-		if c.ID() == id {
-			jq.jobs = append(jq.jobs[:i], jq.jobs[i+1:]...)
-			return nil
+	for i, bucket := range jq.buckets {
+		bucket.mu.Lock()
+		for j, c := range bucket.jobs {
+			if c.ID() == id {
+				jq.buckets[i].jobs = append(bucket.jobs[:j], bucket.jobs[j+1:]...)
+				bucket.mu.Unlock()
+				return nil
+			}
 		}
+		bucket.mu.Unlock()
 	}
 
 	return fmt.Errorf("check with id %s is not in this Job Queue", id)
 }
 
+func (jq *jobQueue) stats() map[string]interface{} {
+	jq.mu.RLock()
+	defer jq.mu.RUnlock()
+
+	nJobs := 0
+	nBuckets := 0
+	for _, bucket := range jq.buckets {
+		nJobs += bucket.size()
+		nBuckets++
+	}
+
+	return map[string]interface{}{
+		"Interval": jq.interval / time.Second,
+		"Buckets":  nBuckets,
+		"Size":     nJobs,
+	}
+}
+
 // run schedules the checks in the queue by posting them to the
 // execution pipeline.
 // Not blocking, runs in a new goroutine.
 func (jq *jobQueue) run(out chan<- check.Check) {
+
+	time.Sleep(time.Second) // wait for one ticker
+
 	go func() {
-		for jq.waitForTick(out) {
+		cases := make([]reflect.SelectCase, 2+len(jq.buckets))
+		cases[oCHAN] = reflect.SelectCase{
+			Dir:  reflect.SelectRecv,
+			Chan: reflect.ValueOf(jq.stop),
+		}
+		cases[hCHAN] = reflect.SelectCase{
+			Dir:  reflect.SelectRecv,
+			Chan: reflect.ValueOf(jq.health.C),
+		}
+
+		ready := true
+		for i, bucket := range jq.buckets {
+
```

Comment:
I am tempted to ask you to wrap `time.Ticker` to allow adding a start time instead of using `AfterFunc`. I really want to kill all the complexity that checks if the tickers have been started to add them to `SelectCase`.
I'll try to find another way if you don't like this one.